!!!TASK: Reduce complexity of ReflectionService

kitsunet opened the pull request:
This change tackles some problems within the reflection service that stem from historically increasing complexity due to various caching mechanisms depending on application context and compile time status.

The aim was to cut down on this complexity, while ensuring that all existing use-cases continue working as intended.

This ultimately also fixes issue #3402 by providing the same reflection data across all possible contexts.

A few features and caches got deprecated with this change and could be breaking in the rare case you used the freeze package api in your code:

The entire concept of freezing a package is deprecated

What remains are the commands in the package controller, which are now all no-ops and deprecated to be removed with 9.0. This is to ensure deployment pipelines possibly calling freeze commands do not break with the 8.4 update.

Additionally the single method `PackageManager::isPackageFrozen` remains, while the rest was removed. None of the methods was ever api and it seems unlikely that someone used them in user-land code. `isPackageFrozen` however is at the very least used in Framework and Neos code and therefore remains until 9.0, but will now return false for every package.

Caches deprecated and unused

With the simplification two caches are no longer needed, both are still declared so that possibly existing cache configuration in user projects doesn't error, but both

`Flow_Reflection_Status`

and

`Flow_Reflection_CompiletimeData`

will no longer be used and any content can be removed.

The only reflection cache is now `Flow_Reflection_RuntimeData`, which makes the name somewhat deceptive as it is also used in compile time. To avoid backwards compatibility issues however it makes sense to keep the name for the foreseeable future.

Quick performance comparisons suggest that especially the initial compile from empty cache benefits from this change. Reflection updates in Development context afterwards seem to be on par with the existing code base.

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

kitsunet committed:
!!!TASK: Reduce complexity of ReflectionService

This change tackles some problems within the reflection service that
stem from historically increasing complexity due to various caching
mechanisms depending on application context and compile time status.

The aim was to cut down on this complexity, while ensuring that all
existing use-cases continue working as intended.

This ultimately also fixes issue #3402 by providing the same
reflection data across all possible contexts.

A few features and caches got deprecated with this change and
could be breaking in the rare case you used the freeze package
api in your code:

The entire concept of freezing a package is deprecated

What remains are the commands in the package controller, which are now
all no-ops and deprecated to be removed with 9.0. This is to ensure
deployment pipelines possibly calling freeze commands do not break
with the 8.4 update.

Additionally the single method `PackageManager::isPackageFrozen`
remains, while the rest was removed. None of the methods was ever api
and it seems unlikely that someone used them in user-land code.
`isPackageFrozen` however is at the very least used in Framework and
Neos code and therefore remains until 9.0, but will now return false
for every package.

Caches deprecated and unused

With the simplification two caches are no longer needed, both are
still declared so that possibly existing cache configuration in user
projects doesn't error, but both

`Flow_Reflection_Status`

and

`Flow_Reflection_CompiletimeData`

will no longer be used and any content can be removed.

The only reflection cache is now `Flow_Reflection_RuntimeData`, which
makes the name somewhat deceptive as it is also used in compile time.
To avoid backwards compatibility issues however it makes sense to keep
the name for the foreseeable future.

Quick performance comparisons suggest that especially the initial
compile from empty cache benefits from this change. Reflection updates
in Development context afterwards seem to be on par with the existing
code base.

kitsunet commented:
Future steps here IMHO could/should be to transfer the huge array structure of reflection data into DTOs and possibly split code that creates the reflection into a Compiletime ReflectionService as we shouldn't need to reflect anything during runtime, therefore making the code easier to reason with. But I resisted from adding those changes here to avoid making it a needlessly large change for 8.4

mhsdesign commented on Neos.Flow/Classes/Command/PackageCommandController.php:
```diff
@@ -165,47 +123,11 @@ public function listCommand(bool $loadingOrder = false)
-
-        foreach ($packagesToFreeze as $packageKey) {
-            $this->packageManager->freezePackage($packageKey);
-            $this->outputLine('Froze package "%s".', [$packageKey]);
-        }
+        $this->outputLine('Package freezing is no longer supported, this command is deprecated and will be removed with 9.0.');
```
juhuu thanks for removing that ❤️

mhsdesign commented on Neos.Flow/Classes/Reflection/ReflectionService.php:
```diff
@@ -762,7 +690,7 @@ public function hasMethod(string $className, string $methodName): bool
-     * @api
+     * @deprecated since 8.4
      */
     public function getMethodTagsValues(string $className, string $methodName): array
```
i suppose this is getting deprecated because its unused system wide?

kitsunet replied:
Yep, correct, there is a test for it, but otherwise it's completely unused.

mhsdesign reviewed (commented):
thank you sooo much for the change! Ill yet have to actually test it and also fully understand the new reflection logic but from reading it briefly i love it. I remember attempting to fix things already in the reflection but the whole initialisation processes is just soo hard to wrap your head around and every-time i got a grasp later i forgot again. Because of that could you please just shortly describe what the expected difference in cache behaviour/performance during Development/Testing and Production might be? Anyway all this freezing functionally will be finally a thing of the past and it definitely does not justify any of the complexity! So super glad to have it gone.

kitsunet commented:
So what this thing do **before**:

`buildReflectionData` is the starting point for actually doing reflection and is triggered via the CompiletimeObjectManager (only). The whole system is a bit disconnected and might be another candidate for future refactoring. 
`buildReflectionData` gets all known classes from the CompiletimeObjectManager, then goes ahead to  call`forgetChangedClasses` which checks if any of the classes no longer exist in the status cache (now removed). Those cache entries get removed from the cach _directly_ in the `CacheManager` based on possibly modifiedfiles found by the `FileMonitor`. This obviously has to happen beforehand. It would be nicer if these things were handled closer together (eg. filemonitor provides the changes to the ReflectionService or so TBD). 
Afterwards `reflectEmergedClasses` will go through the classes and reflect those not reflected.

Then the second big part is the `saveToCache` method called via signal.
This would do a bunch of different things, `updateReflectionData` would write one big cache entry to the reflection compiletime cache (now removed) if there was changed reflection data, BUT only if two other conditions didn't match, one was ($applicationContext = PRODUCTION && runtimecache frozen), no need to write anything because caches were frozen, the other was ("loadFromClassSchemaRuntimeCache" true), which indirectly was the same condition as above decided in `initialize`. IMHO that second condition should never happen as otherwise we should've early returned at the first already. 

Additionally in production context we would write the runtime cache (the only one now) and freeze it afterwards, BUT this production cache did not contain all data the compiletime cache contained leading to bug #3402 
The runtime cache contains separate entries for all caches unlike the compiletime cache which is one big file, this is important later to understand what I did.
Additionally it stores an array of all known classes as an entry (also important later)

And finally in development we would write an additional big cache file outside of any cache backend, that woudl contain about the same as the compiletime cache. This was called `freezePackageReflection` even though it has nothing to do with packages nor with package freezing or cache freezing. It was simply a separate cache file, but this is gone now too.

`initialize` did then decide which of these many caches should be used.

What happens **WITH THIS CHANGE**:

We will discuss this in a slightly different order to make more sense.
`saveToCache` checks if there is updated reflection data and if so runs  `updateCacheEntries` (renamed from `saveProductionData`), which basically works as before BUT stores all necessary data now, this uses the reflection runtime and schemata caches. It will no longer freeze caches.

`reflectEmergedClasses` is still the entry point for (re)generating caches and the flow is as before, we get all known classes from the compiletimeobjectmanager, then we `forgetChangedClasses` which no longer uses the status cache but directly checks the runtime class for the class entries, this is also now the cache were entries get removed from via the CacheManager. But this cuts out the status cache, which was necessary because in compiletime we had that big single file cache entry so we couldn't remove entries from that, so the status cache was a workaround to do this. 

Finally `reflectEmergedClasses` pretty much works as before, with the added special that we prefill the reflectionData with "empty" keys for all known classes, coming from the addtional cache entry `__classNames` in our runtime cache, which is then reduced in `forgetChangedClsases` if necessray, to have a complete list of currently reflected classes to compare the new list to.

`initialize` now always loads the "compounded" cache entries from the runtime cache (so `__classNames`, `__annotatedClasses`, and `__classesByMethodAnnotations`') and class reflection is loaded as needed via `loadOrReflectClassIfNecessary` which is pretty much used everywhere now to make sure we prefer the cache entries.




The class schemata weren't mentioned before, they are basically a sideline and contain data for doctrine ORM retrieved via reflection, so it's kinda in the right place but kinda not. Might also warrant another refactoring, but another day.

kitsunet commented:
Note I might have found two little things to imrpove / fix, shouldn't stop you from reviewing, but should be done before merge.

kitsunet committed:
Make sure interface implementations are cleared

The new caching loads lazily we need to make sure that the interfaces
are actually loaded when trying to  remove implementation classes.

kitsunet committed:
First check if cache updates are scheduled before initialize

Otherwise the initialize might happen for no good reason if the
ReflectionService was never used.

kitsunet committed:
Docblock and code improvements

Some docblock changes and minimal code improvements, might be nice to
have for future refactorings.

kitsunet commented:
Here we go, split in separate commits to  make it easier to review. the last commit is a lot of foobar adaptions to  docblocks and code style, nothing functional.

kdambekalns commented on Neos.Flow/Classes/Package/PackageManager.php:
```diff
@@ -229,6 +228,7 @@ public function getAvailablePackages(): array
      * precompiled reflection data in order to improve performance.
      *
      * @return array<PackageInterface>
+     * @deprecated since 8.4
```
Why is this code still "doing things" instead of being removed or simply returning an empty array?

kitsunet replied:
You are right, total oversight, this should just return [];

kdambekalns commented on Neos.Flow/Classes/Package/PackageManager.php:
```diff
@@ -403,6 +403,7 @@ public function createPackage($packageKey, array $manifest = [], $packagesPath =
      * @param string $toAbsolutePath
      * @return void
      * @throws FilesException
+     * @deprecated since 8.4
```
I think that breaks horribly if actually used these days, no? 🙈

kitsunet replied:
I am pretty sure it does LOL but I didn't want to do too much in this change, deprecating seemed sensible thouhg.

kdambekalns replied:
Absolutely. I was just thinking "good this doesn't constantlyexplode into everyones faces…" 🤣

mhsdesign replied:
haha jup idk how i never found out or ignored move package 😂

kdambekalns commented on Neos.Flow/Configuration/Caches.yaml:
```diff
@@ -93,16 +93,23 @@ Flow_Persistence_Doctrine_Results:
+# Flow_Reflection
+#
+#
+Flow_Reflection_RuntimeData:
+  backend: Neos\Cache\Backend\SimpleFileBackend
+Flow_Reflection_RuntimeClassSchemata:
+  backend: Neos\Cache\Backend\SimpleFileBackend
```
Why do these get an explicit backend configured now?

kitsunet replied:
Because they don't use any features the (more complex) FileBackend offers, and can therefore just use the SimpleFileBackend (should have been the case since ever).

kdambekalns commented on Neos.Flow/Classes/Reflection/ReflectionService.php:
```diff
@@ -1159,9 +1229,10 @@ public function reflectClassProperty(string $className, PropertyReflection $prop
+        $visibility = $this->extractVisibility($property);
         $this->classReflectionData[$className][self::DATA_CLASS_PROPERTIES][$propertyName][self::DATA_PROPERTY_VISIBILITY] = $visibility;
 
+
```
```suggestion
```

kdambekalns commented on Neos.Flow/Classes/Reflection/ReflectionService.php:
```diff
@@ -1727,29 +1825,15 @@ protected function convertParameterReflectionToArray(ParameterReflection $parame
-
         foreach ($classNames as $className) {
-            if (!$this->statusCache->has($this->produceCacheIdentifierFromClassName($className))) {
+            if (is_string($className) && !$this->reflectionDataRuntimeCache->has($this->produceCacheIdentifierFromClassName($className))) {
```
What – if not a string – can `$className` be here?

kitsunet replied:
I had an example were that was _for some reason I couldn't figure out_ a bool, so I jjust wanted to be sure here....

mhsdesign replied:
i think that must have been in a previous iteration of the code as `array_keys` is used which can NEVER return a bool as that is not a valid array key.

kdambekalns commented on Neos.Flow/Tests/Functional/Reflection/ReflectionServiceTest.php:
```diff
@@ -57,6 +57,7 @@ public function classSchemaCanBeBuiltForAggregateRootsWithPlainOldPhpBaseClasses
     /**
      * @test
      * @throws
+     * @deprecated
```
```suggestion
     * @deprecated since 8.4
```

kdambekalns reviewed (approved):
Had a close look, left some comments. Generally fine, so 👍 already.

kitsunet committed:
Remove getFrozenPackages implementation

The method was deprecated and the concept gutted,
this would return empty array anyways.

kitsunet committed:
Apply suggestions from code review

Co-authored-by: Karsten Dambekalns <karsten@dambekalns.de>

kitsunet committed:
Update Neos.Flow/Classes/Reflection/ReflectionService.php

Co-authored-by: Karsten Dambekalns <karsten@dambekalns.de>

mhsdesign commented on Neos.Flow/Classes/Reflection/ReflectionService.php:
```diff
@@ -17,29 +17,28 @@
 use Psr\Log\LoggerInterface;
 use Psr\Log\LogLevel;
 use ReflectionException;
+use ReflectionIntersectionType;
+use ReflectionNamedType;
+use ReflectionType;
+use ReflectionUnionType;
+use RuntimeException;
```
we aggreed that we dont want to import first level things and use `\` - will fix that

mhsdesign replied:
Fixed via 82452e8bd1121949f963b8efd625a9f86866d8f4

mhsdesign committed:
TASK: Global php space symbols should be used fully qualified via backslash

As discussed in https://neos-project.slack.com/archives/C04Q0TC15/p1683558218323329
And voted (8 vs 5) https://neos-project.slack.com/archives/C04Q0TC15/p1683558099679779

We want to use FQN instead

mhsdesign reviewed (approved):
Okay now im wondering whats with the `FreezableBackendInterface` - do we want to deprecate that as well?

Also two times the word "frozen" still appears in the inline docs of the reflection service. Probably the whole doc block of how it works is outdated - but maybe we can put in there youre nice outline from above?

Tested it by upmerging into 9.0 (my 8.4 setup is broken^^) Anyways there are only little merge conflict and besides an error at start (which is expected) and cache flushing the Neos Demo runs fine.
> Warning: unserialize(): Extra data starting at offset 283 of 307 bytes in Framework/Neos.Cache/Classes/Frontend/VariableFrontend.php line 94

I also prepared a followup for 9.0 to remove the dead apis https://github.com/neos/flow-development-collection/pull/3446
and a neos change to remove the usages as well: https://github.com/neos/neos-development-collection/pull/5467

mhsdesign commented on Neos.Flow/Classes/Reflection/ReflectionService.php:
```diff
@@ -1012,14 +1067,17 @@ protected function prepareClassReflectionForUsage(string $className): string
-        $classNamesToReflect = [];
-        foreach ($this->availableClassNames as $classNamesInOnePackage) {
-            $classNamesToReflect = array_merge($classNamesToReflect, $classNamesInOnePackage);
+        $availableClassnames = [];
+        foreach ($this->availableClassNames as $classNamesInPackage) {
+            $availableClassnames[] = $classNamesInPackage;
         }
+        $classNamesToReflect = array_merge([], ...$availableClassnames);
```
okay that part got me first confused ^^ clever though :D

it first looked like a `Object.assign({}, copyFrom)` in javascript but its way better, we unpack all arrays and then merge them together to a flat thing

we can make that a little more simple even:

```suggestion
        // flatten nested array structure to a list of classes
        $availableClassnames = array_merge(...array_values($this->availableClassNames));
```

We dont need the `[]` as since 7.4.0 this function can be called without any parameter, and it will return empty array.

And we still need the `array_values` to prevent: Uncaught ArgumentCountError: array_merge() does not accept unknown named parameters

mhsdesign commented on Neos.Flow/Classes/Reflection/ReflectionService.php:
```diff
@@ -1873,219 +1901,75 @@ protected function loadClassReflectionCompiletimeCache(): bool
         }
+
         if (!$this->initialized) {
             $this->initialize();
```
initialising here when shutting down doesnt make sense and can never be the case i think - but was so beforehand as well ...

kitsunet commented:
FreezableBackendInterface yep it's also for shit like this, that said, can be deprecated but didn't really feel like the scope here.

mhsdesign commented on Neos.Flow/Classes/Reflection/ReflectionService.php:
```diff
@@ -1092,10 +1151,14 @@ protected function reflectClass(string $className): void
             $this->classReflectionData[$className] = [];
         }
 
+        if (!isset($this->classReflectionData[$className][self::DATA_INTERFACE_IMPLEMENTATIONS]) && $class->isInterface()) {
+            $this->classReflectionData[$className][self::DATA_INTERFACE_IMPLEMENTATIONS] = [];
+        }
```
we dont need that explicitly here but it doesnt hurt either right? should be done automatically in `addImplementedInterface`

kitsunet commented:
> Warning: unserialize(): Extra data starting at offset 283 of 307 bytes in Framework/Neos.Cache/Classes/Frontend/VariableFrontend.php line 94

Right that's because of the change to simplefilebackend, which doesn't expect stuff after hte payload, but it's alos faster therefore.

mhsdesign reviewed (commented):
just some other things i found when looking at the code - id still also like the inline docs to be fixed as i wrote above ^^ -> will make a followup pr to tackle my code findings ;) 


Edit: https://github.com/neos/flow-development-collection/pull/3447

mhsdesign commented:
Ha funny i found the original refactoring that introduced this production vs development distinction :D

https://github.com/neos/flow-development-collection/commit/ba1f5fb2adc2a274ce230c2d8328735f7d4e0e8b


And i did the least of performance testing one could do in dev mode and ran `time ./flow flow:core:compile` from cold caches and once after a file change and i could not see any immediate performance drawback before or after
